Breeding-Insight/brapi-Java-TestServer
# BI-2374 Create Endpoints for Processing Batch Deletes and Single Deletes of Lists, Trials, and Samples

# Description
**Stories:** [BI-2374](https://breedinginsight.atlassian.net/jira/software/c/projects/BI/boards/1?selectedIssue=BI-2374), [BI-2378](https://breedinginsight.atlassian.net/jira/software/c/projects/BI/boards/1?selectedIssue=BI-2378), [BI-2371](https://breedinginsight.atlassian.net/jira/software/c/projects/BI/boards/1?selectedIssue=BI-2371), [BI-2375](https://breedinginsight.atlassian.net/jira/software/c/projects/BI/boards/1?selectedIssue=BI-2375)

This feature enhances the BrAPI Java Test Server with new delete operations for Lists, Trials, and Samples entities, introduces a new Batch entity, and implements a soft delete mechanism. Here's a comprehensive summary of the changes:

## Delete Operations

- Added endpoints for deleting single instances of Lists, Trials, and Samples by their respective IDs.
- Implemented batch delete functionality for Lists, Trials, and Samples, allowing for deletion of multiple entities in a single request.
- Introduced soft delete capability for all delete operations:
  - A boolean property `soft_deleted` is added to relevant database tables.
  - Soft-deleted entities are retained in the database but marked as deleted.
  - GET requests and searches exclude soft-deleted entities from results.
  - Hard delete option is still available for permanent removal of entities.
  - Query parameter `hardDelete` (true/false) controls delete behavior, defaulting to soft delete if not specified.

## Batch Entity

- Introduced a new Batch entity to the data model, closely modeled after the existing List entity.
- Created corresponding API endpoints for Batch operations, including creation, retrieval, updating, deletion, item addition, and searching.

## Code Changes

- Implemented `BatchesApi` interface with various methods for Batch operations.
- Added new model classes for Batch-related operations (e.g., `BatchBaseFields`, `BatchDetails`, `BatchNewRequest`, etc.).
- Created `BatchesApiController` to handle Batch-related requests.
- Implemented `BatchService` for business logic related to Batch operations.
- Added `BatchEntity` and `BatchItemEntity` for database persistence.
- Created `BatchRepository` for database interactions.
- Implemented a new factory class for creating components used in processing batch delete requests.

## Database Changes

- Added a new migration script (`V002.002__create_batch_table.sql`) to create the necessary database table for Batch entities.
- Created additional migrations to add the `soft_deleted` column to existing database tables.
- Implemented cascade delete operations for maintaining data integrity.

## Enhancements to Existing Endpoints

- Updated GET and search endpoints to exclude soft-deleted entities from results.
- Modified delete endpoints to support both soft and hard deletes based on the `hardDelete` query parameter.

These changes significantly enhance our API's capabilities for managing Lists, Trials, and Samples, introduce a flexible Batch system, and provide a more nuanced approach to entity deletion with the soft delete mechanism. The new factory class for batch delete components improves the modularity and maintainability of our delete operations.

# Dependencies
**BI-API:** `hackathon/delete-germplasm`

# Testing
Use the Deltabreed UI to do the following for Germplasm Lists, Experiments,and Gentoypic Samples:

1. import at least three entities to be deleted in a later step
2. get the phylo token from an authorized request header Cookie and assign it to the Cookie header for requests made by your API client
3. Get the BrAPI dbIds for the test entitties using any convenient means: for example, `GET http://<base-url>/v1/programs/<program-deltabreed-id>/brapi/v2/lists/` (or /`trials` or `/samples`)
4. soft-delete a single entity by dbId using an API client to send a `DELETE` request to `http://<base-url>/v1/programs/<program-deltabreed-id>/brapi/v2/lists/<BrAPI-entity-dbId>?hardDelete=false`. For at least one of the times this step is carried out, omit the query parameter. Confirm that this request results in a soft delete by default.
5. Confirm on the Deltabreed UI or with the API client that the soft-deleted entity is no longer returned by the BrAPI service. (NOTE: An API client will need to be used to attempt fetching deleted samples. If using the Deltabreed UI to view experiments, be sure to update the cache after deleting a trial.)
6. Verify that the database table for the entity still has a record for that dbID but the value of `soft_deleted` is true.
7. Re-send the `DELETE` request for the same dbId only this time set `hardDelete=true`. Confirm that the record was cleared from the database.
8. Create a batch of entities by sending `POST http://<base-url>/v1/programs/<program-deltabreed-id>/brapi/v2/batches` that includes in the json body the batch type(`lists`,`trials`, or `samples`) and search parameters appropriate to match the remaining test entities:
```
{
  "batchType": "lists",
  "search": {
    "listType": "germplasm",
    "listDbIds": [
      "c4bdfdb4-a6c4-4a88-9f6a-cc133626184f",
      "4b47dc55-8f06-4bb0-a932-fdd01d579bbb",
      "0c289b92-dfae-4332-9011-bb8d7a5a260b"
    ]
  }
}
```
9. Confirm that the response includes the expected entities and copy the batchDbId.
10. Add a new dbId to the batch by sending `POST http://<base-url>/v1/programs/<program-deltabreed-id>/brapi/v2/batches/<batch-DbId>` with json body:
```
[
  "foo-dbId"
]
```
11. View the current state of the batch contents by sending `GET http://<base-url>/v1/programs/<program-deltabreed-id>/brapi/v2/batches/<batch-DbId>` and verifying the data in the response contains `foo-dbId`.
12. Remove `foo-dbId` by sending `PUT http://<base-url>/v1/programs/<program-deltabreed-id>/brapi/v2/batches/<batch-DbId>` with the json body having the desired dbIds:
```
{
    "additionalInfo": null,
    "dateCreated": "2024-10-27T19:51:15-04:00",
    "dateModified": "2024-10-27T19:51:15-04:00",
    "externalReferences": [],
    "batchDescription": null,
    "batchName": null,
    "batchOwnerName": null,
    "batchOwnerPersonDbId": null,
    "batchSize": 3,
    "batchSource": null,
    "batchType": "lists",
    "batchDbId": "d7995a96-65f8-4bc5-880d-f58be3950824",
    "data": [
      "c4bdfdb4-a6c4-4a88-9f6a-cc133626184f",
      "4b47dc55-8f06-4bb0-a932-fdd01d579bbb",
      "0c289b92-dfae-4332-9011-bb8d7a5a260b"
    ]
}
```
13. Confirm the state of the batch is restored as expected by repeating step 11.
14. Delete the batch with `DELETE http://<base-url>/v1/programs/<program-deltabreed-id>/brapi/v2/batches/<batch-dbId>?hardDelete=false`. Confirm that the entities are no longer viewable by Deltabreed UI.
15. When repeating step 14 for another entity type, omit the query parameter and confirm that the entities are still soft-deleted.
16. Repeat step 14 with `hardDelete=true` and confirm that all entities are cleared from the database.

# Checklist:


[BI-2374]: https://breedinginsight.atlassian.net/browse/BI-2374?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
[BI-2378]: https://breedinginsight.atlassian.net/browse/BI-2378?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
[BI-2371]: https://breedinginsight.atlassian.net/browse/BI-2371?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
[BI-2375]: https://breedinginsight.atlassian.net/browse/BI-2375?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
